### PR TITLE
Mana

### PR DESCRIPTION
**Pull request by MemeWizard**
#### Changelog:
- Added mana, mana regeneration and prowess bonuses from unspent mana.
- Added more effects for sorcerer traits
- Changed localisation of non-inheritable sorcerer traits

### CODE REVIEW

File: common/modifiers/wc_mana_modifiers.txt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+﻿wc_prowess_from_availible_mana_modifier = {
```
Comment (TheGroxEmpire) [inline common/modifiers/wc_mana_modifiers.txt:1]: Should probably fix this typo if it doesn't break too much things
```suggestion
﻿wc_prowess_from_available_mana_modifier = {
```

File: common/scripted_effects/wc_mana_effects.txt [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+wc_recalculate_mana_effect = {
+	if = {
+		limit = { var:wc_current_mana_variable > wc_max_mana_value }
+
+		set_variable = { name = wc_current_mana_variable value = wc_max_mana_value }
+	}
+}
+
+wc_recalculate_prowess_from_mana_effect = {
+	remove_all_character_modifier_instances = wc_prowess_from_availible_mana_modifier
```
Comment (TheGroxEmpire) [inline common/scripted_effects/wc_mana_effects.txt:22]:
```suggestion
	remove_all_character_modifier_instances = wc_prowess_from_available_mana_modifier
```

File: common/scripted_effects/wc_mana_effects.txt [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+	while = {
+		count = wc_current_mana_to_prowess_value
+
+		add_character_modifier = { modifier = wc_prowess_from_availible_mana_modifier }
```
Comment (TheGroxEmpire) [inline common/scripted_effects/wc_mana_effects.txt:27]:
```suggestion
		add_character_modifier = { modifier = wc_prowess_from_available_mana_modifier }
```

File: localization/english/wc_mana_l_english.yml [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+ CURRENT_MANA_AMOUNT_LINE:0 "[Character.MakeScope.Var('wc_current_mana_variable').GetValue|0]"
+ MAXIMUM_MANA_AMOUNT_LINE:0 "[GuiScope.SetRoot(Character.MakeScope).ScriptValue('wc_max_mana_value')|0]"
+
+ HUD_TOP_RIGHT_RESOURCES_MAXIMUM_MANA_AMOUNT_LINE:0 "/$MAXIMUM_MANA_AMOUNT_LINE$"
+
+ CHARACTER_WINDOW_MANA_AMOUNT_LINE:0 "$CURRENT_MANA_AMOUNT_LINE$"
+
+ RESOURCES_MANA_BREAKDOWN_HEADER_LINE:0 "Mana: [Character.MakeScope.Var('wc_current_mana_variable').GetValue|0V]/[GuiScope.SetRoot(Character.MakeScope).ScriptValue('wc_max_mana_value')|0V]"
+ RESOURCES_MANA_BREAKDOWN_REGENERATION_LINE:0 "[GuiScope.SetRoot(Character.MakeScope).ScriptValue('wc_mana_regeneration_value')|0+=]/year"
+ RESOURCES_MANA_MODIFIERS_HEADER_LINE:0 "Modifiers from [Character.MakeScope.Var('wc_current_mana_variable').GetValue|0V] availible mana:"
```
Comment (TheGroxEmpire) [inline localization/english/wc_mana_l_english.yml:11]:
```suggestion
 RESOURCES_MANA_MODIFIERS_HEADER_LINE:0 "Modifiers from [Character.MakeScope.Var('wc_current_mana_variable').GetValue|0V] available mana:"
```

File: localization/english/wc_mana_l_english.yml [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+ RESOURCES_MANA_PLAYER_BREAKDOWN_PROWESS_LINE:0 "[prowess|E]:@skill_prowess_icon![GuiScope.SetRoot(Character.MakeScope).ScriptValue('wc_current_mana_to_prowess_value')|0+=] from [GuiScope.SetRoot(Character.MakeScope).ScriptValue('wc_current_mana_to_prowess_percent')|%0V]"
+ RESOURCES_MANA_BREAKDOWN_PERCENTAGE_LINE:0 "Availible: [GuiScope.SetRoot(Character.MakeScope).ScriptValue('wc_current_mana_percent_value')|%V] out of [GuiScope.SetRoot(Character.MakeScope).ScriptValue('wc_max_mana_value')|0]"
```
Comment (TheGroxEmpire) [inline localization/english/wc_mana_l_english.yml:13]:
```suggestion
 RESOURCES_MANA_BREAKDOWN_PERCENTAGE_LINE:0 "Available: [GuiScope.SetRoot(Character.MakeScope).ScriptValue('wc_current_mana_percent_value')|%V] out of [GuiScope.SetRoot(Character.MakeScope).ScriptValue('wc_max_mana_value')|0]"
```

File: localization/english/wc_mana_l_english.yml [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+ wc_prowess_from_availible_mana_modifier:0 "Magical Prowess"
```
Comment (TheGroxEmpire) [inline localization/english/wc_mana_l_english.yml:15]:
```suggestion
 wc_prowess_from_available_mana_modifier:0 "Magical Prowess"
```

**Review by TheGroxEmpire — COMMENTED** [review]
- I think the number beside the percentage here is unnecessary:
<details>

image
</details>

This could be more clear, like "+20 Prowess from 60% of mana pool / current mana"
<details>

image
</details>

Could probably shorten this to "Monthly magic lifestyle experience: +30%":
<details>

image
</details>

- We need game concept for mana explaining what it is, where does it come from, and what it is for. Like what Paradox did to Prowess here is a good example:
<details>

image
</details>

**Comment by MemeWizard** [comment]
> I think the number beside the percentage here is unnecessary:

Idk, it just doesn't feel right when it's not here

<details>
<summary>Click to expand</summary>

image

</details>


> This could be more clear, like "+20 Prowess from 60% of mana pool / current mana"

Done

<details>
<summary>Click to expand</summary>

image

</details>

> Could probably shorten this to "Monthly magic lifestyle experience: +30%":

Can't

> We need game concept for mana explaining what it is, where does it come from, and what it is for. Like what Paradox did to Prowess here is a good example:

Will do (later)

**Comment by MemeWizard** [comment]
> We need game concept for mana explaining what it is, where does it come from, and what it is for. Like what Paradox did to Prowess here is a good example:

Done as far as I can do on this branch.

<details>
<summary>Click to expand</summary>

image
image

</details>

**Review by TheGroxEmpire — APPROVED** [review]
It's all considered then.

**Review by zumbak04 — CHANGES_REQUESTED** [review]

File: localization/english/wc_traits_l_english.yml [hunk 7]
```diff
@@ -469,21 +469,21 @@
  trait_creature_ancient_wind_desc:0 "Ancients of wind are able to tap into the more feral side of nature. Ancients of wind avoid combat but like other ancients, when enemies threaten nature, they battle with their allies."
  trait_creature_ancient_wind_character_desc:0 "$trait_creature_ancient_wind_desc$"
 
- trait_magic_good:0 "Potent Mages"
+ trait_magic_good:0 "Potent Sorcerers"
```
[the remark this answers is not in the document]
Comment (MemeWizard) [reply localization/english/wc_traits_l_english.yml:472]: Yes, just like any other trait category localization
Check `Strengthen Bloodline` decision for examples

File: common/game_concepts/wc_magic_concepts.txt [hunk 8]
```diff
@@ -0,0 +1,5 @@
+﻿wc_mana = {
```
Comment (zumbak04) [inline common/game_concepts/wc_magic_concepts.txt:1]: You should add some debug actions to spend the mana, so we can test your icons.

**Review by zumbak04 — CHANGES_REQUESTED** [review]

File: localization/english/modifiers/wc_modifiers_mana_l_english.yml [hunk 9]
```diff
@@ -0,0 +1,13 @@
+﻿l_english:
+ mana_limit_modifier_line:0 "Maximum [wc_mana|E]: [wc_empty_mana_icon_gc|E]"
+ mana_regeneration_modifier_line:0 "[wc_mana|E] Replenishment: [wc_half_mana_icon_gc|E]"
+ available_mana_to_prowess_modifier_line:0 "Available [wc_mana|E] [wc_full_mana_icon_gc|E] to [prowess|E] @skill_prowess_icon!:"
+
+ TRAIT_FLAG_DESC_wc_magic_good_1_flag:0 "$mana_limit_modifier_line$ [EmptyScope.ScriptValue('wc_sorcerer_trait_1_mana_value')|0=+]\n$mana_regeneration_modifier_line$ [EmptyScope.ScriptValue('wc_sorcerer_trait_1_mana_regeneration_value')|0=+]\n$available_mana_to_prowess_modifier_line$ [EmptyScope.ScriptValue('wc_sorcerer_trait_1_current_mana_to_prowess_value')|0%=+]"
```
[the remark this answers is not in the document]
Comment (MemeWizard) [reply localization/english/modifiers/wc_modifiers_mana_l_english.yml:6]: ???

<details>
<summary>Click to expand</summary>

image

</details>

File: common/game_concepts/wc_magic_concepts.txt (hunk 8 above)
Comment (MemeWizard) [reply common/game_concepts/wc_magic_concepts.txt:1]: `add_%magic%_magic_lifestyle_xp`

[the remark this answers is not in the document]
Comment (MemeWizard) [reply common/game_concepts/wc_magic_concepts.txt:1]: Changed

<details>
<summary>Click to expand</summary>

image

</details>

File: common/on_action/birthday.txt [hunk 10]
```diff
@@ -5,6 +5,31 @@ on_birthday = {
...
+		delay = { months = 7 }
+		wc_on_mana_regeneration = yes
+		delay = { months = 8 }
+		wc_on_mana_regeneration = yes
+		delay = { months = 9 }
+		wc_on_mana_regeneration = yes
+		delay = { months = 10 }
+		wc_on_mana_regeneration = yes
+		delay = { months = 11 }
+		wc_on_mana_regeneration = yes
```
[the remark this answers is not in the document]
Comment (MemeWizard) [reply common/on_action/birthday.txt]: Changed in 917c3f1

File: localization/english/modifiers/wc_modifiers_mana_l_english.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (zumbak04) [reply localization/english/modifiers/wc_modifiers_mana_l_english.yml:6]: An error on my end, most likely caused by a not-loaded loc file.

File: common/game_concepts/wc_magic_concepts.txt (hunk 8 above)
Comment (zumbak04) [reply common/game_concepts/wc_magic_concepts.txt:1]: > Did you meant progressbars?
Yes.

> `add_%magic%_magic_lifestyle_xp`
So there only is a way to increase max mana?

File: common/on_action/wc_mana_on_actions.txt [hunk 11]
```diff
@@ -0,0 +1,6 @@
+﻿wc_on_mana_regeneration = {
```
Comment (zumbak04) [inline common/on_action/wc_mana_on_actions.txt:1]: I didn't notice mana regeneration happening.
<details>
<summary>Click to expand</summary>

image
image

</details>

File: events/wc_events/wc_race_events.txt [hunk 12]
```diff
@@ -513,5 +513,10 @@ WCRAC.5 = {
 			# To avoid recursion
 			# death = { death_reason = death_vanished }
 		}
+
+		# Starts mana regeneration loop
+		trigger_event = {
+			on_action = wc_on_mana_regeneration days = { 1 27 } # So it will not cause a major lag on every 1st of each month
```
Comment (zumbak04) [inline events/wc_events/wc_race_events.txt:519]: If the lag is a point, it's probably will be better if the pulse (`wc_on_mana_regeneration`) is started on the birthday.

Also, it will be super good if we can fire the pulse only when the character picks perks or traits that give you mana. But so far, I don't see such a possibility, so it's sad.

File: gui/hud.gui [hunk 13]
```diff
@@ -78,7 +78,10 @@ widget = {
 	notifications_alerts = {
```
Comment (zumbak04) [inline gui/hud.gui:78]: So you see there's a small gap between the hud and the mana tooltip.
If your mouse goes there, the mana tooltip closes.
<details>
<summary>Click to expand</summary>

image

</details>

**Review by zumbak04 — CHANGES_REQUESTED** [review]

File: gui/hud.gui (hunk 13 above)
Comment (MemeWizard) [reply gui/hud.gui:78]: Report to paradox? It's vanilla interface.

<details>
<summary>Click to expand</summary>

image

</details>

File: common/game_concepts/wc_magic_concepts.txt (hunk 8 above)
Comment (MemeWizard) [reply common/game_concepts/wc_magic_concepts.txt:1]: Use `explorer` console command then

File: gui/hud.gui (hunk 13 above)
Comment (zumbak04) [reply gui/hud.gui:78]: > Report to paradox? It's a part of vanilla interface.

Mana is further to the left than the gold.

File: common/game_concepts/wc_magic_concepts.txt (hunk 8 above)
Comment (zumbak04) [reply common/game_concepts/wc_magic_concepts.txt:1]: > Use `explorer` console command to edit character variables?

But why you don't create debug action? It's faster to click than console commands.

File: gui/hud.gui (hunk 13 above)
Comment (MemeWizard) [reply gui/hud.gui:78]: > Mana is further to the left than the gold.

It uses the same interface as the gold

Comment (zumbak04) [reply gui/hud.gui:78]: > It uses the same interface as the gold

Mana is further to the left than the gold. I compared it.
<details>
<summary>Click to expand</summary>

image
image

</details>

File: common/game_concepts/wc_magic_concepts.txt (hunk 8 above)
Comment (MemeWizard) [reply common/game_concepts/wc_magic_concepts.txt:1]: now `wild_goose_chase_decision` sets your mana to zero

File: gui/hud.gui (hunk 13 above)
Comment (MemeWizard) [reply gui/hud.gui:78]: Better? 79c6fb1dc

File: common/on_action/wc_mana_on_actions.txt (hunk 11 above)
Comment (MemeWizard) [reply common/on_action/wc_mana_on_actions.txt:1]: Should be fixed now

File: localization/english/wc_game_concepts_magic_l_english.yml [hunk 14]
```diff
@@ -0,0 +1,6 @@
+﻿l_english:
+ game_concept_wc_mana:0 "Mana"
+ game_concept_wc_mana_desc:0 "Mana is a resource that represents a Character's ability to bend reality to their will.\nIt can be spend on... currently nothing. Change that and this line later."
```
Comment (zumbak04) [inline localization/english/wc_game_concepts_magic_l_english.yml:3]: I would mention that unspent Mana goes to prowess. It's important to mention for the understanding of this.
<details>
<summary>Click to expand</summary>

image

</details>

File: common/scripted_effects/wc_race_effects.txt [hunk 15]
```diff
@@ -177,6 +177,16 @@ set_race_effect = {
 		limit = { scope:temp_race = flag:creature_kvaldir }
 		become_undead_effect = yes
 	}
+
+	if = { # Starts mana regeneration loop
+		limit = { NOT = { has_variable = potential_friend } }
```
Comment (zumbak04) [inline common/scripted_effects/wc_race_effects.txt:182]: Huh? That's a vanilla term. What it's doing here?
<details>
<summary>Click to expand</summary>

image

</details>

File: common/scripted_effects/wc_race_effects.txt [hunk 16]
```diff
@@ -177,6 +177,16 @@ set_race_effect = {
...
+
+		trigger_event = {
+			on_action = wc_on_mana_regeneration days = { 1 27 } # So it will not cause a major lag on every 1st of each month
+		}
+		set_variable = { name = wc_character_has_mana value = yes }
```
Comment (zumbak04) [inline common/scripted_effects/wc_race_effects.txt:187]: It's not used. Thus may become a leftover at some point.

File: common/scripted_effects/wc_race_effects.txt [hunk 17]
```diff
@@ -177,6 +177,16 @@ set_race_effect = {
 		limit = { scope:temp_race = flag:creature_kvaldir }
 		become_undead_effect = yes
 	}
+
+	if = { # Starts mana regeneration loop
```
Comment (zumbak04) [inline common/scripted_effects/wc_race_effects.txt:181]: I though you made it on the trait/perk adding and that's why you need my review.

**Review by zumbak04 — CHANGES_REQUESTED** [review]

File: common/scripted_effects/wc_race_effects.txt (hunk 15 above)
Comment (MemeWizard) [reply common/scripted_effects/wc_race_effects.txt:182]: weird, don't remember it being there

Comment (MemeWizard) [reply common/scripted_effects/wc_race_effects.txt:182]: Probably forgot to replace it with variable below

Comment (MemeWizard) [inline common/scripted_effects/wc_race_effects.txt:182]:
```suggestion
		limit = { NOT = { has_variable = wc_character_has_mana } }
```

File: common/scripted_effects/wc_race_effects.txt (hunk 17 above)
Comment (MemeWizard) [reply common/scripted_effects/wc_race_effects.txt:181]: As far as I know, it's impossible to add it on perk receiving.

File: localization/english/wc_game_concepts_magic_l_english.yml (hunk 14 above)
Comment (MemeWizard) [reply localization/english/wc_game_concepts_magic_l_english.yml:3]: Done?

File: common/scripted_effects/wc_race_effects.txt (hunk 17 above)
Comment (MemeWizard) [reply common/scripted_effects/wc_race_effects.txt]: I tried

**Review by zumbak04 — CHANGES_REQUESTED** [review]

File: common/game_concepts/wc_magic_concepts.txt [hunk 18]
```diff
@@ -0,0 +1,11 @@
...
```
[the remark this answers is not in the document]
Comment (MemeWizard) [reply common/game_concepts/wc_magic_concepts.txt:1]: Done?

File: common/scripted_effects/wc_mana_effects.txt [hunk 19]
```diff
@@ -0,0 +1,62 @@
+﻿wc_regenerate_mana_effect = {
+	if = {
+		limit = { has_variable = wc_current_mana_variable }
+
+		change_variable = { name = wc_current_mana_variable add = wc_mana_regeneration_value }
+		wc_recalculate_mana_effect = yes
+	}
+	else = { set_variable = { name = wc_current_mana_variable value = wc_max_mana_value } }
+
+	wc_recalculate_prowess_from_mana_effect = yes
```
Comment (zumbak04) [inline common/scripted_effects/wc_mana_effects.txt:10]:
```suggestion
	wc_restore_AMOUNT_of_mana = { AMOUNT = wc_mana_regeneration_value }
```

File: common/scripted_effects/wc_mana_effects.txt [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+		trigger_event = {
+			on_action = wc_on_mana_regeneration
+			days = { 1 28 } # So it will not cause a major lag on every 1st of each month
+		}
+	}
+}
+wc_fill_mana_on_day_one_from_perks_crutch_effect = { # Crutch for mages to have full mana on day 1
+	if = {
+		limit = { game_start_day_trigger = yes }
+		set_variable = { name = wc_current_mana_variable value = wc_max_mana_value }
```
Comment (zumbak04) [inline common/scripted_effects/wc_mana_effects.txt:60]: You set the mana without `wc_recalculate_mana_effect` and `wc_recalculate_prowess_from_mana_effect`.
The most right and clear way to fix it will be creating change current mana and set current mana effects that will handle prowess change, overflows, etc.
The overflow should be fixed by `clamp_variable`.

**Review by zumbak04 — CHANGES_REQUESTED** [review]

File: common/scripted_effects/wc_mana_effects.txt (hunk 19 above)
Comment (zumbak04) [reply common/scripted_effects/wc_mana_effects.txt:10]: Should be fixed during https://github.com/Warcraft-GoA-Development-Team/Warcraft-Guardians-of-Azeroth-2/pull/162#discussion_r664780542.

File: common/scripted_effects/wc_mana_effects.txt (hunk 20 above)
Comment (MemeWizard) [reply common/scripted_effects/wc_mana_effects.txt]: > You set the mana without `wc_recalculate_mana_effect` and `wc_recalculate_prowess_from_mana_effect`.

Fixed?

> The overflow should be fixed by `clamp_variable`.

I tried, but it didn't fix it.

File: common/script_values/wc_mana_values.txt [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+	if = {
+		limit = { has_trait = magic_good_noninheritable }
+		add = wc_sorcerer_trait_noninheritable_mana_value
+	}
+	else_if = {
+		limit = { has_trait = magic_bad_noninheritable }
+		subtract = wc_sorcerer_trait_noninheritable_mana_value
+	}
+
+	ceiling = yes
```
Comment (zumbak04) [inline common/script_values/wc_mana_values.txt:95]: You subtract so to avoid it going negative.
```suggestion
	min = 0
	ceiling = yes
```

File: common/scripted_effects/00_courtier_guest_management_effects.txt [hunk 22]
```diff
@@ -943,6 +943,7 @@ pool_character_development_effect = {
 						}
 					}
 					add_trait = magic_good_noninheritable
+					wc_start_mana_regeneration_loop_effect = yes # Starts mana regeneration loop
```
Comment (zumbak04) [inline common/scripted_effects/00_courtier_guest_management_effects.txt:946]: I would add `add_magical_trait_effect`.

**Review by zumbak04 — CHANGES_REQUESTED** [review]

File: common/script_values/wc_mana_values.txt (hunk 21 above)
Comment (MemeWizard) [reply common/script_values/wc_mana_values.txt:95]: done

File: common/scripted_effects/00_courtier_guest_management_effects.txt (hunk 22 above)
Comment (MemeWizard) [reply common/scripted_effects/00_courtier_guest_management_effects.txt:946]: added

**Review by zumbak04 — APPROVED** [review]
LGTM! Good job!

**Comment by MemeWizard** [comment]
@zumbak04 @TheGroxEmpire thanks for reviews <3
